lib/fizzy/execute.cpp

```diff
@@ -140,14 +140,15 @@ inline uint32_t grow_memory(
 {
     const auto cur_pages = memory.size() / PageSize;
     assert(cur_pages <= size_t(std::numeric_limits<int32_t>::max()));
-    const auto new_pages = cur_pages + delta_pages;
+    const auto new_pages = uint64_t{cur_pages} + delta_pages;
     assert(new_pages >= cur_pages);
     if (new_pages > memory_pages_limit)
         return static_cast<uint32_t>(-1);
 
     try
     {
-        memory.resize(new_pages * PageSize);
+        assert(static_cast<size_t>(new_pages * PageSize) > new_pages);
```

Comment:
Not sure what this checks. 32-bit overflow?

Reply:
Yes, I'll add a test where this fails, and I think `memory_pages_limit` should be capped for 32-bit during instantiate.

This assertion should probably be better changed to testing assumptions about `memory_pages_limit`.

Reply:
Then probably `assert(static_cast<size_t>(new_pages * PageSize) == (new_pages * PageSize))` is better.

Isn't memory capped to 4GB?

Reply:
I checked in the spec, and yes, it is, but it looks like we impose this limit only in parser for memory limits definition in module. It's possible to set `memory_pages_limit` higher, then `memory.grow` will allow it to grow above.

Reply:
Removed this assertion in favor of checking `memory_pages_limit < 4GB`. If that one is true, this must not overflow.